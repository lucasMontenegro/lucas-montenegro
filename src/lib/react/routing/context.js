import React, { createContext, useMemo, useContext } from "react"
import Router from "./Router"
import useClientLinks from "./useClientLinks"
import useTranslationLinks from "./useTranslationLinks"
export const RoutingContext = createContext({
  routing: null,
  route: {
    render: {},
    redirect: null,
  },
})
RoutingContext.displayName = `RoutingContext`
export function RoutingProvider (props) {
  const { routing, location } = props
  const router = useMemo(() => new Router(routing), [routing])
  const route = useMemo(() => router.findRoute(location), [router, location])
  const clientLinks = useClientLinks(props.clientLinks, routing, route)
  const translationLinks = useTranslationLinks(routing, route)
  const value = useMemo(() => (
    { routing, route, clientLinks, translationLinks }
  ), [routing, route, clientLinks, translationLinks])
  return <RoutingContext.Provider value={value}>{props.children}</RoutingContext.Provider>
}
export function useRoute () {
  return useContext(RoutingContext).route
}
export function useRoutingContext () {
  return useContext(RoutingContext)
}