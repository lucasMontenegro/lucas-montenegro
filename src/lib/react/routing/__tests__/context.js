import React, { createContext, useContext } from "react"
import { RoutingContext, RoutingProvider, useRoute, useRoutingContext } from "../context"
import renderer from "react-test-renderer"
jest.mock(`react`, () => {
  const React = jest.requireActual(`react`)
  return {
    ...React,
    __esModule: true,
    default: React,
    createContext: jest.fn(() => ({
      Provider: props => <div {...props} className="RoutingContextProvider" />,
    })),
    useMemo: fn => fn(),
    useContext: jest.fn(),
  }
})
jest.mock(`../Router`, () => ({
  __esModule: true,
  default: class Router {
    findRoute () {
      return { clientName: `foo`, render: { foo: true }, location: { pathname: `/foo` } }
    }
  },
}))
jest.mock(`../useClientLinks`, () => ({
  __esModule: true,
  default: () => [
    {
      active: true,
      clientName: `foo`,
      location: { pathname: `/foo` },
      render: {
        text: { en: () => `en foo`, es: () => `es foo` },
        Icon: () => `foo icon`,
      },
    },
  ],
}))
jest.mock(`../useTranslationLinks`, () => ({
  __esModule: true,
  default: () => ({
    links: [
      { languageCode: `es`, text: `spanish`, location: { pathname: `/foo/es` } },
    ],
    get () {},
  }),
}))
describe(`../context`, () => {
  it(`should use the right versions of its dependencies`, () => {
    expect(jestUtils.getDependencies([`react`])).toMatchSnapshot()
  })
  describe(`RoutingContext`, () => {
    it(`should be configured properly`, () => {
      expect(createContext.mock.calls).toMatchSnapshot()
      createContext.mockClear()
      expect(RoutingContext).toMatchSnapshot()
    })
  })
  describe(`RoutingProvider`, () => {
    it(`should render`, () => {
      expect(renderer.create(
        <RoutingProvider routing={{ clientNames: [`foo`], languageCodes: [`en`, `es`] }}>
          children
        </RoutingProvider>
      ).toJSON()).toMatchSnapshot()
    })
  })
  describe(`useRoute`, () => {
    it(`should return the route object from the context`, () => {
      const route = {}
      useContext.mockReturnValueOnce({ route })
      expect(useRoute()).toBe(route)
      expect(useContext.mock.calls[0][0]).toBe(RoutingContext)
      useContext.mockClear()
    })
  })
  describe(`useRoute`, () => {
    it(`should return the context value`, () => {
      const context = {}
      useContext.mockReturnValueOnce(context)
      expect(useRoutingContext()).toBe(context)
      expect(useContext.mock.calls[0][0]).toBe(RoutingContext)
      useContext.mockClear()
    })
  })
})