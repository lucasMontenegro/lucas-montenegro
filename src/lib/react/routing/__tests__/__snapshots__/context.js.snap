// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`../context RoutingContext should be configured properly 1`] = `
Array [
  Array [
    Object {
      "route": Object {
        "redirect": null,
        "render": Object {},
      },
      "routing": null,
    },
  ],
]
`;

exports[`../context RoutingContext should be configured properly 2`] = `
Object {
  "Provider": [Function],
  "displayName": "RoutingContext",
}
`;

exports[`../context RoutingProvider should render 1`] = `
<div
  className="RoutingContextProvider"
  value={
    Object {
      "clientLinks": Array [
        Object {
          "active": true,
          "clientName": "foo",
          "location": Object {
            "pathname": "/foo",
          },
          "render": Object {
            "Icon": [Function],
            "text": Object {
              "en": [Function],
              "es": [Function],
            },
          },
        },
      ],
      "route": Object {
        "clientName": "foo",
        "location": Object {
          "pathname": "/foo",
        },
        "render": Object {
          "foo": true,
        },
      },
      "routing": Object {
        "clientNames": Array [
          "foo",
        ],
        "languageCodes": Array [
          "en",
          "es",
        ],
      },
      "translationLinks": Object {
        "get": [Function],
        "links": Array [
          Object {
            "languageCode": "es",
            "location": Object {
              "pathname": "/foo/es",
            },
            "text": "spanish",
          },
        ],
      },
    }
  }
>
  children
</div>
`;

exports[`../context should use the right versions of its dependencies 1`] = `
Object {
  "react": "^16.10.1",
}
`;
